### Segmenting and Clustering Neighborhoods in Toronto

The idea of this activity is to identify the neighborhoods in Toronto that are similar in terms of venues (restaurants, gyms, clubs...). First, a Wikipedia page containing Toronto postal codes, boroughs and neighborhoods is scrapped in order to obtain that information. Then, latitude and the longitude coordinates of each neighborhood is obtained using Geocoder Python package. For last, Foursqare API is used to retrieve the closest venues to the center of each neighbothood and clustering is applied to group those neighborhoods based on the venues similarity.

#### 1. Getting Toronto's postal codes, boroughs and neighborhoods

In [1]:
import pandas as pd
!conda install -c conda-forge lxml --yes

table = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')
table = table[0]

#Ignoring cells with a borough that is Not assigned.
postalcodes = table[table.Borough != 'Not assigned'].reset_index(drop=True)
postalcodes.head()

Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

Although the description of the exercise says that more than one neighborhood can exist in one postal code area, this is not correct. Apparently Wikipedia has already merged neighborhoods based on postal code:

In [2]:
postalcodes[postalcodes['Postal Code'].duplicated()]

Empty DataFrame
Columns: [Postal Code, Borough, Neighborhood]
Index: []

The description also says that "If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough", but all the boroughs not "Not Assigned" have a neighborhood assigned: 

In [3]:
postalcodes['Neighborhood'].isnull().sum()

0

In [4]:
postalcodes.rename(columns={'Postal Code': 'PostalCode'}, inplace=True)
postalcodes.shape

(103, 3)

### 2. Getting the latitude and the longitude coordinates of each neighborhood

It was suggested the use of geocoder to this task. However, the following code got stucked in the while loop meaning that geocoder was not able to return a result.

In [5]:
#!conda install -c conda-forge geocoder --yes
#import geocoder

#coords_df = pd.DataFrame(columns=['Latitude', 'Longitude'])

#for i in range(1, postalcodes.size):
    # initialize coords variable to None
    #lat_lng_coords = None

    # loop until getting the coordinates since sometimes geocoder return nothing and it is necessary to persist
    #while(lat_lng_coords is None):
     #   g = geocoder.google('{}, Toronto, Ontario'.format(postalcodes.loc[i,'PostalCode']))
      #  lat_lng_coords = g.latlng

    #coords_df = coords_df.append({'Latitude':lat_lng_coords[0], 'Longitude':lat_lng_coords[1]}, ignore_index=True)
    
#coords_df

In case it is not able to get the geographical coordinates of the neighborhoods using the Geocoder package, a csv file that has the geographical coordinates of each postal code is provided.

In [6]:
toronto_coords = pd.read_csv('https://cocl.us/Geospatial_data')
toronto_coords.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

The coordinates of each postal code is searched in the data provided by the csv...

In [7]:
coords_df = pd.DataFrame(columns=['Latitude', 'Longitude'])

for i in range(0, len(postalcodes)):
    row = toronto_coords[toronto_coords.loc[:,'Postal Code'] == postalcodes.loc[i, 'PostalCode']]
    coords_df = coords_df.append({'Latitude':row.iloc[0]['Latitude'], 'Longitude':row.iloc[0]['Longitude']}, ignore_index=True)
    
coords_df.head()

Latitude  Longitude
0  43.753259 -79.329656
1  43.725882 -79.315572
2  43.654260 -79.360636
3  43.718518 -79.464763
4  43.662301 -79.389494

... and combined with the previous data to obtain the following data frame:

In [8]:
toronto_neighborhoods = pd.concat([postalcodes, coords_df], axis=1)
toronto_neighborhoods

PostalCode           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
..         ...               ...   
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

Analysing better the Neighbohoods, it was found that there are duplicates:

In [9]:
toronto_neighborhoods[toronto_neighborhoods['Neighborhood'].duplicated()]

PostalCode     Borough Neighborhood   Latitude  Longitude
13        M3C  North York    Don Mills  43.725900 -79.340923
46        M3L  North York    Downsview  43.739015 -79.506944
53        M3M  North York    Downsview  43.728496 -79.495697
60        M3N  North York    Downsview  43.761631 -79.520999

Let's change their names to differentiate them:

In [11]:
toronto_neighborhoods.loc[13, 'Neighborhood'] = toronto_neighborhoods.loc[13, 'Neighborhood'] + ' 2'
toronto_neighborhoods.loc[46, 'Neighborhood'] = toronto_neighborhoods.loc[46, 'Neighborhood'] + ' 2'
toronto_neighborhoods.loc[53, 'Neighborhood'] = toronto_neighborhoods.loc[53, 'Neighborhood'] + ' 3'
toronto_neighborhoods.loc[60, 'Neighborhood'] = toronto_neighborhoods.loc[60, 'Neighborhood'] + ' 4'
toronto_neighborhoods

PostalCode           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
..         ...               ...   
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

### 3. Finding similar neighborhoods based on venues similarity

Using the Foursquare API to get the top venues close to each neighborhood:

In [12]:
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Foursquare API credentials and settings
CLIENT_ID = 'TJNIF0UICB1EMN1LETJN0YNM5DARNO3JKGK4ZJ41RT4QRT5Q' # your Foursquare ID
CLIENT_SECRET = 'RGNFFRIQEWPAQLRLTDOXQVEIISJQWRFBOTSOJWLHDBAVHVWS' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100

# Function that gets the top nearby venues for a list of coordinates (lat, long)
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):          
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Running the above function on each neighborhood:

In [13]:
toronto_venues = getNearbyVenues(names=toronto_neighborhoods['Neighborhood'],
                                   latitudes=toronto_neighborhoods['Latitude'],
                                   longitudes=toronto_neighborhoods['Longitude']
                                  )
toronto_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2         Parkwoods              43.753259              -79.329656   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude     Venue Category  
0         Brookbanks Park       43.751976       -79.332140               Park  
1             649 Variety       43.754513       -79.331942  Convenience Store  
2           Variety Store       43.751974       -79.333114  Food & Drink Shop  
3  Victoria Village Arena       43.723481       -79.315635       Hockey Arena  
4             Tim Hortons       43.725517       -79.313103        Coffee Shop

Transforming each venue category in dummy variable:

In [14]:
# one hot encoding
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column
toronto_onehot.drop(['Neighborhood'], axis=1, inplace=True)
toronto_onehot = pd.concat([toronto_venues['Neighborhood'], toronto_onehot], axis=1)

toronto_onehot.head()

Neighborhood  Accessories Store  Afghan Restaurant  Airport  \
0         Parkwoods                  0                  0        0   
1         Parkwoods                  0                  0        0   
2         Parkwoods                  0                  0        0   
3  Victoria Village                  0                  0        0   
4  Victoria Village                  0                  0        0   

   Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0                   0               0                0                 0   
1                   0               0                0                 0   
2                   0               0                0                 0   
3                   0               0                0                 0   
4                   0               0                0                 0   

   American Restaurant  Antique Shop  Aquarium  Art Gallery  Art Museum  \
0                    0             0         0            0           0   
1                    0             0         0            0           0   
2                    0             0         0            0           0   
3                    0             0         0            0           0   
4                    0             0         0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Garage  \
0                    0                 0                   0            0   
1                    0                 0                   0            0   
2                    0                 0                   0            0   
3                    0                 0                   0            0   
4                    0                 0                   0            0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          0           0           0       0     0    0   
1              0          0           0           0       0     0    0   
2              0          0           0           0       0     0    0   
3              0          0           0           0       0     0    0   
4              0          0           0           0       0     0    0   

   Baseball Field  Baseball Stadium  Basketball Court  Basketball Stadium  \
0               0                 0                 0                   0   
1               0                 0                 0                   0   
2               0                 0                 0                   0   
3               0                 0                 0                   0   
4               0                 0                 0                   0   

   Beach  Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  \
0      0                0         0           0                   0   
1      0                0         0           0                   0   
2      0                0         0           0                   0   
3      0                0         0           0                   0   
4      0                0         0           0                   0   

   Bike Shop  Bistro  Boat or Ferry  Bookstore  Boutique  \
0          0       0              0          0         0   
1          0       0              0          0         0   
2          0       0              0          0         0   
3          0       0              0          0         0   
4          0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                     0               0        0            0   
1                     0               0        0            0   
2                     0               0        0            0   
3                     0               0        0            0   
4                     0               0        0            0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0                0         0             0              0

Grouping rows by neighborhood and taking the mean of the frequency of occurrence of each category

In [15]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Accessories Store  \
0                                           Agincourt           0.000000   
1                              Alderwood, Long Branch           0.000000   
2     Bathurst Manor, Wilson Heights, Downsview North           0.000000   
3                                     Bayview Village           0.000000   
4                   Bedford Park, Lawrence Manor East           0.000000   
5                                         Berczy Park           0.000000   
6                         Birch Cliff, Cliffside West           0.000000   
7        Brockton, Parkdale Village, Exhibition Place           0.000000   
8   Business reply mail Processing Centre, South C...           0.000000   
9   CN Tower, King and Spadina, Railway Lands, Har...           0.000000   
10                                Caledonia-Fairbanks           0.000000   
11              Canada Post Gateway Processing Centre           0.000000   
12                                          Cedarbrae           0.000000   
13                                 Central Bay Street           0.000000   
14                                           Christie           0.000000   
15                               Church and Wellesley           0.000000   
16            Clarks Corners, Tam O'Shanter, Sullivan           0.000000   
17    Cliffside, Cliffcrest, Scarborough Village West           0.000000   
18                     Commerce Court, Victoria Hotel           0.000000   
19                                         Davisville           0.000000   
20                                   Davisville North           0.000000   
21   Del Ray, Mount Dennis, Keelsdale and Silverthorn           0.000000   
22                                          Don Mills           0.000000   
23                                        Don Mills 2           0.000000   
24  Dorset Park, Wexford Heights, Scarborough Town...           0.000000   
25                                          Downsview           0.000000   
26                                        Downsview 2           0.000000   
27                                        Downsview 3           0.000000   
28                                        Downsview 4           0.000000   
29                       Dufferin, Dovercourt Village           0.000000   
30      East Toronto, Broadview North (Old East York)           0.000000   
31  Eringate, Bloordale Gardens, Old Burnhamthorpe...           0.000000   
32                       Fairview, Henry Farm, Oriole           0.000000   
33             First Canadian Place, Underground city           0.000000   
34    Forest Hill North & West, Forest Hill Road Park           0.000000   
35                           Garden District, Ryerson           0.000000   
36                                          Glencairn           0.000000   
37                    Golden Mile, Clairlea, Oakridge           0.000000   
38                  Guildwood, Morningside, West Hill           0.000000   
39  Harbourfront East, Union Station, Toronto Islands           0.000000   
40                      High Park, The Junction South           0.000000   
41                                  Hillcrest Village           0.000000   
42                                      Humber Summit           0.000000   
43                                   Humberlea, Emery           0.000000   
44                                 Humewood-Cedarvale           0.000000   
45                     India Bazaar, The Beaches West           0.000000   
46        Kennedy Park, Ionview, East Birchmount Park           0.000000   
47          Kensington Market, Chinatown, Grange Park           0.000000   
48  Kingsview Village, St. Phillips, Martin Grove ...           0.000000   
49                   Lawrence Manor, Lawrence Heights           0.153846   
50                                      Lawrence Park           0.000000   
51                                            Leaside           0.000000   
52           

Creating a new dataframe to represent the top 10 venues for each neighborhood.

In [16]:
import numpy as np
#Sort the venues in decending order
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East            Restaurant   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0        Breakfast Spot  Latin American Restaurant          Skating Rink   
1    Athletics & Sports                   Pharmacy           Coffee Shop   
2                  Bank                   Pharmacy   Fried Chicken Joint   
3                  Bank         Chinese Restaurant   Japanese Restaurant   
4        Sandwich Place         Italian Restaurant           Coffee Shop   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0   Dumpling Restaurant   Distribution Center               Dog Run   
1                  Pool                   Pub        Sandwich Place   
2           Bridal Shop         Shopping Mall                 Diner   
3                 Diner        Discount Store   Distribution Center   
4           Pizza Place               Butcher       Thai Restaurant   

  8th Most Common Venue      9th Most Common Venue 10th Most Common Venue  
0      Doner Restaurant                 Donut Shop              Drugstore  
1          Skating Rink                        Gym           Antique Shop  
2        Sandwich Place  Middle Eastern Restaurant            Supermarket  
3               Dog Run           Doner Restaurant            Yoga Studio  
4             Juice Bar                       Café                    Pub

Using clustering (K-Means) to group the neighborhoods according to the top 10 most commom venues

In [17]:
from sklearn.cluster import KMeans

# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_neighborhoods

# merge neighborhoods_venues_sorted with toronto_neighborhoods to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged

PostalCode           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   
12         M1C       Scarborough   
13         M3C        North York   
14         M4C         East York   
15         M5C  Downtown Toronto   
16         M6C              York   
17         M9C         Etobicoke   
18         M1E       Scarborough   
19         M4E      East Toronto   
20         M5E  Downtown Toronto   
21         M6E              York   
22         M1G       Scarborough   
23         M4G         East York   
24         M5G  Downtown Toronto   
25         M6G  Downtown Toronto   
26         M1H       Scarborough   
27         M2H        North York   
28         M3H        North York   
29         M4H         East York   
30         M5H  Downtown Toronto   
31         M6H      West Toronto   
32         M1J       Scarborough   
33         M2J        North York   
34         M3J        North York   
35         M4J         East York   
36         M5J  Downtown Toronto   
37         M6J      West Toronto   
38         M1K       Scarborough   
39         M2K        North York   
40         M3K        North York   
41         M4K      East Toronto   
42         M5K  Downtown Toronto   
43         M6K      West Toronto   
44         M1L       Scarborough   
45         M2L        North York   
46         M3L        North York   
47         M4L      East Toronto   
48         M5L  Downtown Toronto   
49         M6L        North York   
50         M9L        North York   
51         M1M       Scarborough   
52         M2M        North York   
53         M3M        North York   
54         M4M      East Toronto   
55         M5M        North York   
56         M6M              York   
57         M9M        North York   
58         M1N       Scarborough   
59         M2N        North York   
60         M3N        North York   
61         M4N   Central Toronto   
62         M5N   Central Toronto   
63         M6N              York   
64         M9N              York   
65         M1P       Scarborough   
66         M2P        North York   
67         M4P   Central Toronto   
68         M5P   Central Toronto   
69         M6P      West Toronto   
70         M9P         Etobicoke   
71         M1R       Scarborough   
72         M2R        North York   
73         M4R   Central Toronto   
74         M5R   Central Toronto   
75         M6R      West Toronto   
76         M7R       Mississauga   
77         M9R         Etobicoke   
78         M1S       Scarborough   
79         M4S   Central Toronto   
80         M5S  Downtown Toronto   
81         M6S      West Toronto   
82         M1T       Scarborough   
83         M4T   Central Toronto   
84         M5T  Downtown Toronto   
85         M1V       Scarborough   
86         M4V   Central Toronto   
87         M5V  Downtown Toronto   
88         M8V         Etobicoke   
89         M9V         Etobicoke   
90         M1W       Scarborough   
91         M4W  Downtown Toronto   
92         M5W  Downtown Toronto   
93         M8W         Etobicoke   
94         M9W         Etobicoke   
95         M1X       Scarborough   
96         M4X  Downtown Toronto   
97         M5X  Downtown Toronto   
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  \
0                                            Parkwoods  43.753259 -79.329656   
1                                     Victoria Village  43.725882 -79.315572   
2                  

There are some NaN values because Foursquare returned venues for only 99 of the 103 neighborhoods. Thus, let's remove those records since the goal of this notebook is to find similarity among neighborhoods based on their venues.

In [18]:
toronto_merged.dropna(inplace=True)
toronto_merged.shape

(99, 16)

Since the murge occured between dataframes of diffent size, the merged table presented some NaN values in the column Cluster Labels. As NaN as of type float, the merging process converted the column Cluster Labels to float. Let's put it back to int.

In [19]:
toronto_merged['Cluster Labels'] = toronto_merged['Cluster Labels'].astype(int)

Finally, let's visualize the resulting clusters in a map to check which neighborhoods are similar.

In [20]:
#!conda install -c conda-forge folium=0.5.0 --yes
#!conda install -c conda-forge geopy --yes

import folium # map rendering library
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
import matplotlib.cm as cm
import matplotlib.colors as colors

#getting Toronto coordinates
address = 'Toronto, Ontario'
geolocator = Nominatim(user_agent="explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters